clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java

```diff
@@ -762,16 +777,33 @@
         RESPONSES[ApiKeys.SYNC_GROUP.id] = SYNC_GROUP_RESPONSE;
         RESPONSES[ApiKeys.DESCRIBE_GROUPS.id] = DESCRIBE_GROUPS_RESPONSE;
         RESPONSES[ApiKeys.LIST_GROUPS.id] = LIST_GROUPS_RESPONSE;
+        RESPONSES[ApiKeys.API_VERSION.id] = API_VERSION_RESPONSE;
 
-        /* set the maximum version of each api */
-        for (ApiKeys api : ApiKeys.values())
+        /* set the minimum and maximum version of each api */
+        for (ApiKeys api : ApiKeys.values()) {
```

Comment:
With the versioning getting a little more complex below, I wonder if it makes sense to try and encapsulate some of this logic a little better. For example, rather than using a raw array for each request type, we could have something like this:
```java
class KafkaApi {
  private ApiKey api;
  private Schema[] requests;
  private Schema[] responses;

  Schema currentSchema();
  Schema schemaFor(int version);
  int minVersion();
  int currentVersion();
}
```